packages/flutter/lib/src/rendering/proxy_box.dart

```diff
@@ -1264,8 +1264,43 @@ class RenderCustomPaint extends RenderProxyBox {
   }
 
   void _paintWithPainter(Canvas canvas, Offset offset, CustomPainter painter) {
+    int debugPreviousCanvasSaveCount;
+    assert(() { debugPreviousCanvasSaveCount = canvas.getSaveCount(); return true; });
     canvas.translate(offset.dx, offset.dy);
+    canvas.save(); // this will be essentially free if the custom painter doesn't change the transform/clip/whatever
```

Comment:
We should save before the translate and skip the reverse translation.  The overhead of having 4 Dart->C++ calls rather than 3 Dart->C++ calls dwarfs whatever we might save by eliding the save/restore pair in the SkPicture.